python/lsst/analysis/tools/tasks/sourceObjectTableAnalysis.py

```diff
@@ -98,35 +241,147 @@ class SourceObjectTableAnalysisConfig(
+        """Callback function to be used with reconstructor."""
+        return self.prepareAssociatedSources(
+            dataId["visit"],
+            inputs["data"],
+            inputs["associatedSources"],
+            inputs["refCat"],
+            inputs["visitTable"],
+            inputs["properMotionCatalog"],
+        )
 
-        # Get isolated sources:
-        visit = inputs["data"].dataId["visit"]
-        band = inputs["data"].dataId["band"]
-        names = self.collectInputNames()
-        names -= {self.config.ra_column, self.config.dec_column}
-        data = inputs["data"].get(parameters={"columns": names})
+    def applyProperMotionCorrections(
+        self, isolatedSources, properMotionCatalog, visitTable, visit, refEpochs
+    ):
+        """Shift source positions to match the epoch of the reference catalog
+        objects.
 
-        dataId = butlerQC.quantum.dataId
-        plotInfo = self.parsePlotInfo(inputs, dataId)
+        Parameters
+        ----------
+        isolatedSources : `astropy.table.Table`
+            Catalog of sources to which proper motion corrections will be made.
+        properMotionCatalog : `pd.DataFrame`
+            Catalog with proper motion and parallax information.
+        visitTable : `pd.DataFrame`
+            Catalog containing the epoch for the visit corresponding to the
+            isolatedSources.
+        visit : `int`
+            Identifier of the isolatedSources' visit.
+        """
+        sourceMjd = visitTable.loc[visit]["expMidptMJD"]
+
+        # Get target date from reference catalog
+        targetEpochs = refEpochs.to_numpy()
+        # There may not be a valid reference epoch on the edge of a given
+        # region. Do not make a proper motion correction for any sources on the
+        # edge.
+        targetEpochs[~np.isfinite(targetEpochs)] = sourceMjd
+
+        with Matcher(isolatedSources["coord_ra"], isolatedSources["coord_dec"]) as m:
+            idx, i1, i2, d = m.query_radius(
+                properMotionCatalog["starX"], properMotionCatalog["starY"], 0.2 / 3600, return_indices=True
+            )
+        pmDf = properMotionCatalog.rename(
+            columns={
+                "starX": "ra",
+                "starY": "dec",
+                "starPMx": "pmRA",
+                "starPMy": "pmDec",
+                "starParallax": "parallax",
+            }
+        )
+        pmDf.insert(5, "MJD", sourceMjd)
+        raCorrection, decCorrection = calculate_apparent_motion(pmDf.iloc[i2], targetEpochs[i1])
+        isolatedSources["coord_ra"][i1] -= raCorrection.value
+        isolatedSources["coord_dec"][i1] -= decCorrection.value
+        return isolatedSources
 
+    def prepareAssociatedSources(
+        self, visit, data, associatedSourceRefs, refCats, visitTable, properMotionCatalog
+    ):
+        """Match isolated sources with reference objects and shift the sources
+        to the object epochs if the properMotionCatalog is not None.
```

Comment:
I don't see anything in this method that allows `properMotionCatalog` to be None?

Reply:
Fixed text.